### PR TITLE
Update common channels to almalinux mirror list

### PR DESCRIPTION
**Pull request by jordimassaguerpla**
## What does this PR change?

Partial fix for https://github.com/SUSE/spacewalk/issues/14388

## GUI diff

No difference.

- [ ] **DONE**

## Documentation
- No documentation needed

- [ ] **DONE**

## Test coverage
- No tests

- [ ] **DONE**

## Links

Partial fix for https://github.com/SUSE/spacewalk/issues/14388

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by jordimassaguerpla** [comment]
I followed https://www.uyuni-project.org/uyuni-docs/en/uyuni/client-configuration/clients-almalinux.html#_add_software_channels to add the channels and syncronize and it looks good.

**Comment by jordimassaguerpla** [comment]
I have installed a server from scratch, added the almalinux channels and then applyed the sql script. All looks good.

**Comment by jordimassaguerpla** [comment]
@juliogonzalez this is ready. This needs https://github.com/uyuni-project/uyuni/pull/4324

File: schema/spacewalk/upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/002-use_almalinux_mirror.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+UPDATE rhnContentSource SET source_url = 'https://mirrors.almalinux.org/mirrorlist/8/baseos' where label like 'External - AlmaLinux 8 (%)';
```
Comment (juliogonzalez) [inline schema/spacewalk/upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/002-use_almalinux_mirror.sql:1]: Nitpick:

```suggestion
UPDATE rhnContentSource SET source_url = 'https://mirrors.almalinux.org/mirrorlist/8/baseos' WHERE label LIKE 'External - AlmaLinux 8 (%)';
```
(same for the other lines)

File: utils/spacewalk-common-channels.ini [hunk 2]
```diff
@@ -2917,36 +2917,36 @@ name     = AlmaLinux 8 (%(arch)s)
 gpgkey_url = https://repo.almalinux.org/almalinux/RPM-GPG-KEY-AlmaLinux
 gpgkey_id = 3ABB34F8
 gpgkey_fingerprint = 5E9B 8F56 17B5 066C E920 57C3 488F CF7C 3ABB 34F8
-repo_url = https://repo.almalinux.org/almalinux/8/BaseOS/%(arch)s/os/
+repo_url = https://mirrors.almalinux.org/mirrorlist/8/baseos
```
Comment (juliogonzalez) [inline utils/spacewalk-common-channels.ini:2920]: For coherence, I'd add the trailing slash:

```suggestion
repo_url = https://mirrors.almalinux.org/mirrorlist/8/baseos/
```
Same for the other lines

File: schema/spacewalk/upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/002-use_almalinux_mirror.sql (hunk 1 above)
Comment (juliogonzalez) [reply schema/spacewalk/upgrade/susemanager-schema-4.3.3-to-susemanager-schema-4.3.4/002-use_almalinux_mirror.sql:1]: Can you also do an two extra test for me?

InstanceA, using current master:
- Add all the channels for all architectures, not just `x86_64`
- Apply the SQL
- Try to sync all of them.

Instance B, using this PR:
- Add all the channels for all architectures, not just `x86_64`
- Try to sync all of them.

In the second case, I am 99.99% sure there won't be problems.

In the first one, I am less sure. Should, but... :-D

**Review by juliogonzalez — APPROVED** [review]
PR LGTM. Just some style things, and a request for two extra tests, just in case.

**Comment by jordimassaguerpla** [comment]
@juliogonzalez I cannot add aarch64 arch : 
```
spacewalk-common-channels -a aarch64 almalinux8 almalinux8-uyuni-client almalinux8-appstream
SUSE Manager username: admin
SUSE Manager password: 
No channels matching your selection.
```
This was before applying the sql change and without the changes from this PR.

**Comment by juliogonzalez** [comment]
>

> @juliogonzalez I cannot add aarch64 arch :
> 
> ```
> spacewalk-common-channels -a aarch64 almalinux8 almalinux8-uyuni-client almalinux8-appstream
> SUSE Manager username: admin
> SUSE Manager password: 
> No channels matching your selection.
> ```
> 
> This was before applying the sql change and without the changes from this PR.

Check if you need to rebase from master. I see `aarch64` at https://github.com/uyuni-project/uyuni/blob/master/utils/spacewalk-common-channels.ini#L2914

**Comment by jordimassaguerpla** [comment]
> > 
> 
> > @juliogonzalez I cannot add aarch64 arch :
> > ```
> > spacewalk-common-channels -a aarch64 almalinux8 almalinux8-uyuni-client almalinux8-appstream
> > SUSE Manager username: admin
> > SUSE Manager password: 
> > No channels matching your selection.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This was before applying the sql change and without the changes from this PR.
> 
> Check if you need to rebase from master. I see `aarch64` at https://github.com/uyuni-project/uyuni/blob/master/utils/spacewalk-common-channels.ini#L2914

That was it. I was using stable packages.

Now I have this problem when running the update commands in the database:
```
ERROR:  duplicate key value violates unique constraint "rhn_cs_repo_uq"
DETAIL:  Key (COALESCE(org_id, 0::numeric), type_id, source_url, (
CASE
    WHEN label::text ~~ 'manifest_%'::text THEN 1
    ELSE 0
END))=(1, 500, https://mirrors.almalinux.org/mirrorlist/8/appstream, 0) already exists.
```
Seems repo urls are expected to be unique and now with this mirror thing urls for different archs are the same, and this "vilotes" the index/constraint

**Comment by jordimassaguerpla** [comment]
> rhn_cs_repo_uq

I can drop the index by adding at the top of the script

```
rhn_cs_repo_uq
```

**Comment by jordimassaguerpla** [comment]
@cbosdo Can you look at the last comments? If we use the same URL for different channels for almalinux, this violates an index we have. My question is that, do we need different channels defined for each arch? Or should we have only one entry at the database?

**Comment by juliogonzalez** [comment]
> > > 
> > 
> > 
> > > @juliogonzalez I cannot add aarch64 arch :
> > > ```
> > > spacewalk-common-channels -a aarch64 almalinux8 almalinux8-uyuni-client almalinux8-appstream
> > > SUSE Manager username: admin
> > > SUSE Manager password: 
> > > No channels matching your selection.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This was before applying the sql change and without the changes from this PR.
> > 
> > 
> > Check if you need to rebase from master. I see `aarch64` at https://github.com/uyuni-project/uyuni/blob/master/utils/spacewalk-common-channels.ini#L2914
> 
> That was it. I was using stable packages.
> 
> Now I have this problem when running the update commands in the database:
> 
> ```
> ERROR:  duplicate key value violates unique constraint "rhn_cs_repo_uq"
> DETAIL:  Key (COALESCE(org_id, 0::numeric), type_id, source_url, (
> CASE
>     WHEN label::text ~~ 'manifest_%'::text THEN 1
>     ELSE 0
> END))=(1, 500, https://mirrors.almalinux.org/mirrorlist/8/appstream, 0) already exists.
> ```
> 
> Seems repo urls are expected to be unique and now with this mirror thing urls for different archs are the same, and this "vilotes" the index/constraint

Bingo, this is why I wanted to check this.

You are right about the URLs expected to be unique. `spacewalk-common-channels` is able to handle this by reusing an existing repo for several channels, so if you add both architectures on a new server, it should work (check it as well, please).

But now for the update, you it seems you will need to handle duplicates, leaving only one set of URLs, then adjust all the channels accordingly. You could need some help from developers, at this is outside my area of knowledge.

My suggestion:

1. Pick one of the architectures, assign those repos to all the relevant channels
2. Remove all the repos for the other architectures.

**Comment by juliogonzalez** [comment]
> > rhn_cs_repo_uq
> 
> I can drop the index by adding at the top of the script
> 
> ```
> rhn_cs_repo_uq
> ```

Nope!!!!!! Read the other comment I left :-D

**Comment by cbosdo** [comment]
> @cbosdo Can you look at the last comments? If we use the same URL for different channels for almalinux, this violates an index we have. My question is that, do we need different channels defined for each arch? Or should we have only one entry at the database?

Honestly I don't know. @mcalmer and @lucidd probably know more on that.

**Comment by jordimassaguerpla** [comment]
I have started from scratch and add the almalinux channels using the modified common channels ini file from this PR (/etc/rhn/spacewalk-common-channels.ini).

Then,  I have run this sql command at the database:

```select rhnchannel.label, rhncontentsource.label, source_url from rhnchannelcontentsource, rhncontentsource, rhnchannel where source_id=rhncontentsource.id and channel_id=rhnchannel.id```

 and the result is this:

 ```
                 label                 |                                label                                 |                                                source_url                                                 
---------------------------------------+----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------
 almalinux8-x86_64                     | External - AlmaLinux 8 (x86_64)                                      | https://mirrors.almalinux.org/mirrorlist/8/baseos/
 almalinux8-uyuni-client-devel-x86_64  | External - Uyuni Client Tools for AlmaLinux 8 (x86_64) (Development) | https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/EL8-Uyuni-Client-Tools/EL_8/
 almalinux8-uyuni-client-x86_64        | External - Uyuni Client Tools for AlmaLinux 8 (x86_64)               | https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/
 almalinux8-x86_64-appstream           | External - AlmaLinux 8 AppStream (x86_64)                            | https://mirrors.almalinux.org/mirrorlist/8/appstream/
 almalinux8-x86_64-extras              | External - AlmaLinux 8 Extras (x86_64)                               | https://mirrors.almalinux.org/mirrorlist/8/extras/
 almalinux8-x86_64-ha                  | External - AlmaLinux 8 High Availability (x86_64)                    | https://mirrors.almalinux.org/mirrorlist/8/ha/
 almalinux8-x86_64-powertools          | External - AlmaLinux 8 PowerTools (x86_64)                           | https://mirrors.almalinux.org/mirrorlist/8/powertools/
 almalinux8-aarch64                    | External - AlmaLinux 8 (x86_64)                                      | https://mirrors.almalinux.org/mirrorlist/8/baseos/
 almalinux8-aarch64-appstream          | External - AlmaLinux 8 AppStream (x86_64)                            | https://mirrors.almalinux.org/mirrorlist/8/appstream/
 almalinux8-aarch64-extras             | External - AlmaLinux 8 Extras (x86_64)                               | https://mirrors.almalinux.org/mirrorlist/8/extras/
 almalinux8-aarch64-ha                 | External - AlmaLinux 8 High Availability (x86_64)                    | https://mirrors.almalinux.org/mirrorlist/8/ha/
 almalinux8-aarch64-powertools         | External - AlmaLinux 8 PowerTools (x86_64)                           | https://mirrors.almalinux.org/mirrorlist/8/powertools/
 almalinux8-uyuni-client-aarch64       | External - Uyuni Client Tools for AlmaLinux 8 (x86_64)               | https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/EL8-Uyuni-Client-Tools/EL_8/
 almalinux8-uyuni-client-devel-aarch64 | External - Uyuni Client Tools for AlmaLinux 8 (x86_64) (Development) | https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/EL8-Uyuni-Client-Tools/EL_8/
(14 rows)
```

Thus the sql should update these tables:
```rhnchannelcontentsource, rhncontentsource, rhnchannel```

The first one contains the relation between the last two.

**Comment by jordimassaguerpla** [comment]
Using the .ini file you end up with this contents in the database:

```
channel label                          | content source label                                                   | source_url
---------------------------------+--------------------------------------------------------+-----------------------------------------------------------
 almalinux8-x86_64-ha           | External - AlmaLinux 8 High Availability (x86_64)      | https://mirrors.almalinux.org/mirrorlist/8/ha
 almalinux8-aarch64-ha           | External - AlmaLinux 8 High Availability (x86_64)      | https://mirrors.almalinux.org/mirrorlist/8/ha

```

while with the upgrade script, we get:

```
channel label                          | content source label                                                   | source_url
---------------------------------+--------------------------------------------------------+-----------------------------------------------------------
almalinux8-x86_64               | External - AlmaLinux 8                                 | https://mirrors.almalinux.org/mirrorlist/8/baseos

 almalinux8-aarch64-ha           | External - AlmaLinux 8 High Availability               | https://mirrors.almalinux.org/mirrorlist/8/ha

```
Note the first one has x86_64 in the content source label while the channel label is aarch64.

In the UI you don't see a difference, it is only visible in the database.
My guess is that the code that adds the content source and channels based on .ini file uses the name in there:
```
name     = AlmaLinux 8 AppStream (%(arch)s)
```
I tried using instead
```
name     = AlmaLinux 8 AppStream
```
but then I got an error about duplicating the name in the channel, so that is not an option.

@juliogonzalez This is not seen by the user, it could even be ignored. My only concern is that further sql commands may depend on that label. Actually, if you run the sql script after having added the almalinux channels with the ini file, this gets fixed. So, can we make the user run the update script? Or, do we need to worry at all about this if the thing is working?

**Comment by juliogonzalez** [comment]
> Using the .ini file you end up with this contents in the database:
> 
> ```
> channel label                          | content source label                                                   | source_url
> ---------------------------------+--------------------------------------------------------+-----------------------------------------------------------
>  almalinux8-x86_64-ha           | External - AlmaLinux 8 High Availability (x86_64)      | https://mirrors.almalinux.org/mirrorlist/8/ha
>  almalinux8-aarch64-ha           | External - AlmaLinux 8 High Availability (x86_64)      | https://mirrors.almalinux.org/mirrorlist/8/ha
> ```

Try to change the `.ini` so the `name` does not contain the architecture. `content source label` is only used for repositories, not for channels.

> while with the upgrade script, we get:
> 
> ```
> channel label                          | content source label                                                   | source_url
> ---------------------------------+--------------------------------------------------------+-----------------------------------------------------------
> almalinux8-x86_64               | External - AlmaLinux 8                                 | https://mirrors.almalinux.org/mirrorlist/8/baseos
> 
>  almalinux8-aarch64-ha           | External - AlmaLinux 8 High Availability               | https://mirrors.almalinux.org/mirrorlist/8/ha
> ```
> 
> Note the first one has x86_64 in the content source label while the channel label is aarch64.
> 
> In the UI you don't see a difference, it is only visible in the database. My guess is that the code that adds the content source and channels based on .ini file uses the name in there:

`channel_label`is something you use with CLI as well, for example with `spacewalk-reposync`. `content_source_label`, however is something for "humans" and to create the relationsuip between channels and repos, AFAIK

> ```
> name     = AlmaLinux 8 AppStream (%(arch)s)
> ```
> 
> I tried using instead
> 
> ```
> name     = AlmaLinux 8 AppStream
> ```
> 
> but then I got an error about duplicating the name in the channel, so that is not an option.

You need to care about repositories, and the relationship between channels and repos, but there's no need to change the channels. Those need to be specific for architectures.

> @juliogonzalez This is not seen by the user, it could even be ignored. My only concern is that further sql commands may depend on that label. Actually, if you run the sql script after having added the almalinux channels with the ini file, this gets fixed. So, can we make the user run the update script? Or, do we need to worry at all about this if the thing is working?

Can you have a huddle with mc tomorrow? I think he will clarify the DB structure better than I did on this comment. We need to make sure we do the right things.

File: utils/spacewalk-utils.changes [hunk 3]
```diff
@@ -1,3 +1,4 @@
+- Add almalinux8 mirror list
```
Comment (juliogonzalez) [inline utils/spacewalk-utils.changes:1]:
```suggestion
- Use Almalinux Mirror List
```

**Comment by jordimassaguerpla** [comment]
> > Using the .ini file you end up with this contents in the database:
> > ```
> > channel label                          | content source label                                                   | source_url
> > ---------------------------------+--------------------------------------------------------+-----------------------------------------------------------
> >  almalinux8-x86_64-ha           | External - AlmaLinux 8 High Availability (x86_64)      | https://mirrors.almalinux.org/mirrorlist/8/ha
> >  almalinux8-aarch64-ha           | External - AlmaLinux 8 High Availability (x86_64)      | https://mirrors.almalinux.org/mirrorlist/8/ha
> > ```
> 
> Try to change the `.ini` so the `name` does not contain the architecture. `content source label` is only used for repositories, not for channels.
> 

I  had tried already and no luck, it needs the architecture or it complains about a duplication in the database.

> > while with the upgrade script, we get:
> > ```
> > channel label                          | content source label                                                   | source_url
> > ---------------------------------+--------------------------------------------------------+-----------------------------------------------------------
> > almalinux8-x86_64               | External - AlmaLinux 8                                 | https://mirrors.almalinux.org/mirrorlist/8/baseos
> > 
> >  almalinux8-aarch64-ha           | External - AlmaLinux 8 High Availability               | https://mirrors.almalinux.org/mirrorlist/8/ha
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Note the first one has x86_64 in the content source label while the channel label is aarch64.
> > In the UI you don't see a difference, it is only visible in the database. My guess is that the code that adds the content source and channels based on .ini file uses the name in there:
> 
> `channel_label`is something you use with CLI as well, for example with `spacewalk-reposync`. `content_source_label`, however is something for "humans" and to create the relationsuip between channels and repos, AFAIK
> 

`channel_label` is correct, the issue is only with the `content source label`.

**Comment by jordimassaguerpla** [comment]
> 
> `channel_label` is correct, the issue is only with the `content source label`.

@juliogonzalez I checked with @mcalmer and he told me this is not an issue. The content source label will be different anyway depending on whether the user added first the x86_64 or the aarch64 channels.

@juliogonzalez this PR is ready!

**Pull request merged**